src/client/dfs/dfs.c

```diff
@@ -3500,6 +3500,18 @@ dfs_readdir(dfs_t *dfs, dfs_obj_t *obj, daos_anchor_t *anchor, uint32_t *nr,
 				       kds[i].kd_key_len + 1, "%s", ptr);
 			D_ASSERT(len >= kds[i].kd_key_len);
 			ptr += kds[i].kd_key_len;
+
+			/** stat the entry if requested */
+			if (stbufs) {
+				rc = entry_stat(dfs, DAOS_TX_NONE, obj->oh, dirs[key_nr].d_name,
+						kds[i].kd_key_len, NULL, true, &stbufs[key_nr],
+						NULL);
```

Comment:
This function is inherently racy against different clients so it would be possible to define the API such that a return value indicated an error but if *nr was non-zero then at least that many dirents/stbufs are valid.  The anchor itself would be in a undefined state in that case however so I'm not sure how a user would recover from that case.

Reply:
if i understand your comment, that's the model of the relaxed mode, where these races are acceptable. i would argue that this race is also acceptable in the balanced mode, but in that case we can used a dtx for the readdir and stat. in any case this isn't supported till daos 3.0 anyway.

Reply:
I was thinking if a race was encountered and the stat failed the API would be able to return *some* valid data but it seems a corner case.

Reply:
ok i see what you mean, and i think that's OK. indeed that a user cannot recover against a race error where the stat failed because the anchor and the number of items returned do not match. so either they have to save the initial passed in anchor and restart from that, or restart from the beginning.